x/precisebank/keeper/send.go

```diff
@@ -19,23 +32,298 @@ func (k Keeper) SendCoins(
+
+	// New balance exceeds max fractional balance, so we need to carry it over
+	// to the integer balance.
+	carryRequired := newFractionalBalance.GTE(types.ConversionFactor())
+
+	if carryRequired {
+		// Carry over to integer amount
+		newFractionalBalance = newFractionalBalance.Sub(types.ConversionFactor())
+	}
+
+	return newFractionalBalance, carryRequired
+}
+
+// SendCoinsFromModuleToModule transfers coins from a ModuleAccount to another.
+// It will panic if either module account does not exist. An error is returned
+// if the recipient module is the x/precisebank module account or if sending the
+// tokens fails.
 func (k Keeper) SendCoinsFromAccountToModule(
 	ctx sdk.Context,
 	senderAddr sdk.AccAddress,
 	recipientModule string,
 	amt sdk.Coins,
 ) error {
-	panic("unimplemented")
+	recipientAcc := k.ak.GetModuleAccount(ctx, recipientModule)
+	if recipientAcc == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", recipientModule))
+	}
+
+	if recipientModule == types.ModuleName {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s is not allowed to receive funds", types.ModuleName)
+	}
+
+	return k.SendCoins(ctx, senderAddr, recipientAcc.GetAddress(), amt)
 }
 
+// SendCoinsFromModuleToAccount transfers coins from a ModuleAccount to an AccAddress.
+// It will panic if the module account does not exist. An error is returned if
+// the recipient address is blocked, if the sender is the x/precisebank module
+// account, or if sending the tokens fails.
 func (k Keeper) SendCoinsFromModuleToAccount(
 	ctx sdk.Context,
 	senderModule string,
 	recipientAddr sdk.AccAddress,
 	amt sdk.Coins,
 ) error {
-	panic("unimplemented")
+	// Identical panics to x/bank
+	senderAddr := k.ak.GetModuleAddress(senderModule)
+	if senderAddr == nil {
+		panic(errorsmod.Wrapf(sdkerrors.ErrUnknownAddress, "module account %s does not exist", senderModule))
+	}
+
+	// Custom error to prevent external modules from modifying x/precisebank
+	// balances. x/precisebank module account balance is for internal reserve
+	// use only.
+	if senderModule == types.ModuleName {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "module account %s is not allowed to send funds", types.ModuleName)
+	}
+
+	// Uses x/bank BlockedAddr, no need to modify. x/precisebank should be
+	// blocked.
+	if k.bk.BlockedAddr(recipientAddr) {
+		return errorsmod.Wrapf(sdkerrors.ErrUnauthorized, "%s is not allowed to receive funds", recipientAddr)
+	}
+
+	return k.SendCoins(ctx, senderAddr, recipientAddr, amt)
+}
+
+// updateInsufficientFundsError returns a modified ErrInsufficientFunds with
+// extended coin amounts if the error is due to insufficient funds. Otherwise,
+// it returns the original error. This is used since x/bank transfers will
+// return errors with integer coins, but we want the more accurate error that
+// contains the full extended coin balance and send amounts.
+func (k Keeper) updateInsufficientFundsError(
```

Comment:
nit: can we name this `wrapInsufficientFundsError` for more consistent naming with other error wrappers/updates?

Reply:
hmm I named this to not be wrap since it's not actually wrapping but returning a new error with the correct full balance 🤔 thought it would be a bit misleading to be called wrap but can rename if its still clear enough

Reply:
hm yeah.. haha no strong preference. keep `update` 😆